nixos/modules/services/web-apps/nextcloud-backup.nix

```diff
@@ -0,0 +1,199 @@
+          sudo -u nextcloud ${mysql} ${commonFlags} ${cfg.config.dbname} < $1/nextcloud-backup/${backupFileName}
+        '';
+      }
+    else throw "Unknown nextcloud database type ${cfg.config.dbtype}";
+  checkBackup = ''
+    for f in \
+      "$1/nextcloud-backup/config.php" \
+      "$1/nextcloud-backup/${dbBackupRestore.backupFileName}" \
+      "$1/nextcloud-backup/store-apps" \
+      "$1/nextcloud-backup/data"
+    do
+      if [ ! -e "$f" ] ;then
+        echo "Did not find file $f."
+        echo "It looks like the backup is incomplete, aborting."
+        exit 1
+      fi
+    done
+  '';
+  backupScript = pkgs.writeScriptBin "nextcloud-backup" ''
+    #! ${pkgs.runtimeShell}
+    set -Eeuo pipefail
+    if [ "$#" != "1" ] || [ "$1" = "-h" ] || [ "$1" = "--help" ] ;then
+      echo "Usage: nextcloud-backup TARGET_DIR"
+      echo "The backup will be created in TARGET_DIR/nextcloud-backup."
+      exit 1
+    fi
+    if [ "$(id -un)" != "nextcloud" ] ;then
+      echo "This script has to be run as the nextcloud user, aborting."
+      exit 1
+    fi
+    # See https://docs.nextcloud.com/server/latest/admin_manual/maintenance/backup.html
+    echo "Enabling maintenance mode..."
+    ${cfg.occ}/bin/nextcloud-occ maintenance:mode --on
+    echo "Copying config and manually installed apps..."
+    mkdir -p $1/nextcloud-backup
+    ${backupRsyncCommand} ${cfg.datadir}/config/config.php $1/nextcloud-backup
+    ${backupRsyncCommand} ${cfg.home}/store-apps $1/nextcloud-backup
+    echo "Copying data directory..."
+    ${backupRsyncCommand} ${cfg.datadir}/data $1/nextcloud-backup
+    echo "Dumping database..."
+    ${dbBackupRestore.backupCommand}
+    echo "Disabling maintenance mode..."
+    ${cfg.occ}/bin/nextcloud-occ maintenance:mode --off
+    echo "Nextcloud Backup Metadata" > $1/nextcloud-backup/metadata.txt
+    echo "Backup Time: $(date --rfc-3339=seconds)" >> $1/nextcloud-backup/metadata.txt
+    echo "Nextcloud Version: ${cfg.package.version}" >> $1/nextcloud-backup/metadata.txt
+    echo "Backup Format Version: 1" >> $1/nextcloud-backup/metadata.txt
+    echo "Database Type: ${cfg.config.dbtype}" >> $1/nextcloud-backup/metadata.txt
+    ${checkBackup}
+    echo "Nextcloud backup completed."
+  '';
+  restoreScript = pkgs.writeScriptBin "nextcloud-restore" ''
+    #! ${pkgs.runtimeShell}
+    set -Eeuo pipefail
+    if [ "$#" != "1" ] || [ "$1" = "-h" ] || [ "$1" = "--help" ] ;then
+      echo "Usage: nextcloud-backup BACKUP_DIR"
+      echo "Restore the backup contained in BACKUP_DIR/nextcloud-backup."
+      echo "Note that this requires a working nextcloud installation which will then be overwritten."
+      exit 1
+    fi
+    if [ "$(id -u)" != "0" ] ;then
+      echo "This script has to be run as root, aborting."
+      exit 1
+    fi
+    # Read backup metadata
+    NEXTCLOUD_MAJOR_VERSION=$(grep "^Nextcloud Version:" "$1/nextcloud-backup/metadata.txt" | sed -r "s/Nextcloud Version: ([0-9]+)\..*/\1/")
```

Comment:
Wouldn't it make sense to use some kind of structured data here? Grepping for texts seems a little brittle IMHO.